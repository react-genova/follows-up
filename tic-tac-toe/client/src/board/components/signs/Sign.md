### With values

```jsx
const { SIGN_O, SIGN_X } = require('../../modules/types/signs.constants.js');
<div style={{ display:"flex", width:"120px", height:"50px", padding:"2px", backgroundColor: "rgb(50,50,50)" }}>
    <Sign hasValue type={SIGN_X} />
    <Sign hasValue type={SIGN_O} />
</div>
```

### Without values (hover)

```jsx
const { SIGN_O, SIGN_X } = require('../../modules/types/signs.constants.js');
<div style={{ display:"flex", width:"120px", height:"50px", padding:"2px", backgroundColor: "rgb(50,50,50)" }}>
    <Sign type={SIGN_X} />
    <Sign type={SIGN_O} />
</div>
```

### A little bigger and mixed

```jsx
const { SIGN_O, SIGN_X } = require('../../modules/types/signs.constants.js');
<div style={{ display:"flex", width:"480px", height:"100px", padding:"5px", backgroundColor: "rgb(50,50,50)" }}>
    <Sign hasValue type={SIGN_X} />
    <Sign hasValue type={SIGN_O} />
    <Sign type={SIGN_X} />
    <Sign type={SIGN_O} />
</div>
```