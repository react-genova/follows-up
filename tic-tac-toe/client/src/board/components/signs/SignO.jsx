import React from 'react';
import PropTypes from 'prop-types';

const SignO = ({ ghost }) => (
    <svg width="100%" height="100%" viewBox="0 0 100 100">
        <ellipse cx="50" cy="50" rx="46" ry="46" fill="transparent" stroke={ghost ? 'lightgray' : 'yellow'} strokeWidth="8" />
    </svg>
);

SignO.propTypes = {
    ghost: PropTypes.bool,
};

SignO.defaultProps = {
    ghost: false,
};

export default SignO;
