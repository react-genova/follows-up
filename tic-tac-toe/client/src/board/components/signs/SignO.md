### On light background

```jsx
<div style={{ display:"flex", width:"100px", height:"50px", padding:"2px" }}>
    <SignO />
    <SignO ghost />
</div>
```

### On dark background

```jsx
<div style={{ display:"flex", width:"100px", height:"50px", padding:"2px", backgroundColor: "rgb(50,50,50)" }}>
    <SignO />
    <SignO ghost />
</div>
```

### A little bigger

```jsx
<div style={{ display:"flex", width:"200px", height:"100px", padding:"5px", backgroundColor: "rgb(50,50,50)" }}>
    <SignO />
    <SignO ghost />
</div>
```