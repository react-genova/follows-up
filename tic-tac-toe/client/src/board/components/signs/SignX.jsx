import React from 'react';
import PropTypes from 'prop-types';

const SignX = ({ ghost }) => (
    <svg width="100%" height="100%" viewBox="0 0 100 100">
        <line x1="0" y1="0" x2="100" y2="100" stroke={ghost ? 'lightgray' : 'cyan'} strokeWidth="8" />
        <line x1="0" y1="100" x2="100" y2="0" stroke={ghost ? 'lightgray' : 'cyan'} strokeWidth="8" />
    </svg>
);

SignX.propTypes = {
    ghost: PropTypes.bool,
};

SignX.defaultProps = {
    ghost: false,
};


export default SignX;
