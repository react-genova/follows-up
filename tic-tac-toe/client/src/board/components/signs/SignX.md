### On light background

```jsx
<div style={{ display:"flex", width:"100px", height:"50px", padding:"2px" }}>
    <SignX />
    <SignX ghost />
</div>
```

### On dark background

```jsx
<div style={{ display:"flex", width:"100px", height:"50px", padding:"2px", backgroundColor: "rgb(50,50,50)" }}>
    <SignX />
    <SignX ghost />
</div>
```

### A little bigger

```jsx
<div style={{ display:"flex", width:"200px", height:"100px", padding:"5px", backgroundColor: "rgb(50,50,50)" }}>
    <SignX />
    <SignX ghost />
</div>
```