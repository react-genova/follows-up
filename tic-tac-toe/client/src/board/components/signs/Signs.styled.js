import styled from 'styled-components';

const SignContainer = styled.div`
    width: 100%;
    height: 100%;
`;

export default SignContainer;
