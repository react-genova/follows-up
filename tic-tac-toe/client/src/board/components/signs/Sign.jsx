import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import * as SIGNS from '../../modules/types/signs.constants';
import SignContainer from './Signs.styled';
import SignO from './SignO';
import SignX from './SignX';

class Sign extends PureComponent {
    constructor(props) {
        super(props);

        this.state = {
            hover: false,
        }
    }

    onMouseEnter = () => this.setState({ hover: true });

    onMouseLeave = () => this.setState({ hover: false });

    onClick = () => this.props.onClick(this.props.type);

    render() {
        const { type, hasValue } = this.props;
        const { hover } = this.state;
        return (
            <SignContainer onClick={this.onClick} onMouseEnter={this.onMouseEnter} onMouseLeave={this.onMouseLeave}>
            {
                (hasValue || hover) && {
                    [SIGNS.SIGN_O]: <SignO ghost={!hasValue} />,
                    [SIGNS.SIGN_X]: <SignX ghost={!hasValue} />,
                }[type]
            }
            </SignContainer>
        );
    }
}

Sign.propTypes = {
    hasValue: PropTypes.bool,
    onClick: PropTypes.func,
    type: PropTypes.oneOf([SIGNS.SIGN_O, SIGNS.SIGN_X]),
};

Sign.defaultProps = {
    hasValue: false,
    onClick: () => null,
    type: SIGNS.SIGN_O,
};

export default Sign;
