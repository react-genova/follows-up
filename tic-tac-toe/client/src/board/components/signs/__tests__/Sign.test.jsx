import React from 'react';
import { render, cleanup, fireEvent } from 'react-testing-library';
import 'jest-dom/extend-expect';
import Sign from '../Sign';
import { SIGN_O, SIGN_X } from '../../../modules/types/signs.constants';

describe('SignO component', () => {
    afterEach(cleanup);

    it('renders without crashing', () => {
        const { container } = render(<Sign />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders default sign with value', () => {
        const { container } = render(<Sign hasValue />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders O sign with value', () => {
        const { container } = render(<Sign hasValue type={SIGN_O} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders X sign with value', () => {
        const { container } = render(<Sign hasValue type={SIGN_X} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('invokes click callback', () => {
        const onClick = jest.fn();
        const { container } = render(<Sign className="sign-test" hasValue type={SIGN_X} onClick={onClick} />);
        fireEvent.click(container.querySelector('.sign-test'));
        expect(onClick).toHaveBeenCalledTimes(1);
    });

    it('simulates hover active without value', () => {
        const onClick = jest.fn();
        const { container } = render(<Sign className="sign-test" type={SIGN_X} onClick={onClick} />);
        fireEvent.mouseEnter(container.querySelector('.sign-test'));
        expect(container.firstChild).toMatchSnapshot();
        expect(container.querySelectorAll('.sign-x').length).toBe(1);
        expect(container.querySelectorAll('.sign-o').length).toBe(0);
        fireEvent.mouseLeave(container.querySelector('.sign-test'));
        expect(container.firstChild).toMatchSnapshot();
        expect(container.querySelectorAll('.sign-x').length).toBe(0);
        expect(container.querySelectorAll('.sign-o').length).toBe(0);
        expect(onClick).toHaveBeenCalledTimes(0);
    });
});
