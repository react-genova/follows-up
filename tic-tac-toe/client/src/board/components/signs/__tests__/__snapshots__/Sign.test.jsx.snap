// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignO component renders O sign with value 1`] = `
<div
  class="sc-bdVaJa hWaRAZ"
>
  <svg
    height="100%"
    viewBox="0 0 100 100"
    width="100%"
  >
    <ellipse
      cx="50"
      cy="50"
      fill="transparent"
      rx="46"
      ry="46"
      stroke="yellow"
      stroke-width="8"
    />
  </svg>
</div>
`;

exports[`SignO component renders X sign with value 1`] = `
<div
  class="sc-bdVaJa hWaRAZ"
>
  <svg
    height="100%"
    viewBox="0 0 100 100"
    width="100%"
  >
    <line
      stroke="cyan"
      stroke-width="8"
      x1="0"
      x2="100"
      y1="0"
      y2="100"
    />
    <line
      stroke="cyan"
      stroke-width="8"
      x1="0"
      x2="100"
      y1="100"
      y2="0"
    />
  </svg>
</div>
`;

exports[`SignO component renders default sign with value 1`] = `
<div
  class="sc-bdVaJa hWaRAZ"
>
  <svg
    height="100%"
    viewBox="0 0 100 100"
    width="100%"
  >
    <ellipse
      cx="50"
      cy="50"
      fill="transparent"
      rx="46"
      ry="46"
      stroke="yellow"
      stroke-width="8"
    />
  </svg>
</div>
`;

exports[`SignO component renders without crashing 1`] = `
<div
  class="sc-bdVaJa hWaRAZ"
/>
`;

exports[`SignO component simulates hover active without value 1`] = `
<div
  class="sc-bdVaJa hWaRAZ"
>
  <svg
    height="100%"
    viewBox="0 0 100 100"
    width="100%"
  >
    <line
      stroke="lightgray"
      stroke-width="8"
      x1="0"
      x2="100"
      y1="0"
      y2="100"
    />
    <line
      stroke="lightgray"
      stroke-width="8"
      x1="0"
      x2="100"
      y1="100"
      y2="0"
    />
  </svg>
</div>
`;

exports[`SignO component simulates hover active without value 2`] = `
<div
  class="sc-bdVaJa hWaRAZ"
/>
`;
